### ISSUE TITLE
Document steps need to port cuda only libs (pytorch)

### ISSUE THREAD

**Issue by delijati** [hughperkins/cuda-on-cl#12]
It would be nice to a have step by step documentation on howto convert to/ add support for `opencl` for a third party lib like [pytorch](https://github.com/pytorch/pytorch/).

**Comment by hughperkins** [comment]
Hmmm, thats always a tricky question to answer.

You already have your own repos, so that seems like a good sign :-)

Having said that, just to get this out of the way: there is no easy answer, there is no 'port me' button :-)  And if there was an easy way to port pytorch in 40-80 hours, I'd have probably already done it :-)  For indicative purposes, this was about how many hours I needed for various OpenCL-related things, in the past:

- write DeepCL from scratch: ~5 months * 4.5 weeks per month * 40 hours per week (cos I have a full time job, so after that + sleeping, that only leaves me 40 hours to work on my opensource stuff)
  - = 900 hours
- port enough of luatorch to opencl that people started using it: 6-8 weeks * 40 hours per week
  - ~= 300 hours
- get enough of cuda-on-cl working so tensorflow can run a basic mlp-type network, with matrix multiplications, training ,derivatives, per-element operatinos, reductions: ~8 weeks * 40 hours per week
  - ~= 300 hours
  - (note that almost all of this effort was in cuda-on-cl itself, not in tensorflow, so should be mostly re-usable, but still... juts to give an idea)

First thing I think you'd want to do, if your primary goal is to port pytorch, I reckon is:

- familiarize yourself with pytorch.  Where is the CUDA code?  How easy to factorize the CUDA code out, and put some OpenCL code in its place?  How to go about creating an OpenCL context and so on?
   - for luatorch, the answer to factorization was: very easy :-)  since cuda was in a separate repository, and optional
   - so I just created a new github repo for cltorch, ie http://github.com/hughperkins/cltorch
   - ... and just started hacking:
     - create an OpenCL context
     - port enough stuff across so that can eg create some simple 1-d tensor, with length 1, but with no way to read/write it for now :-P
     - add code to somehow write a value to it
     - add code to somehow read a value from it
     - ^^^ you probably want to at least do this much for pytorch, just simply by hand, to get a feel for how it works
- you might want to look at how cltorch works, probably in parallel with the above.  Maybe not in detail, but at least to have the feeling that you have understanding for how it inserts itself into torch, integrates with torch etc
- once you've got a feel for how to start adding opencl to pytorch, I think you have a few choices, but I would think you'd want to start by evaluating the options available to you:
   - try doing some simple computecpp program, presuambly in sycl. build it, compile it, run it.  how well does it work?  good points?  bad points?
  - ditto for trisycl
  - ditto for cuda-on-cl
  - (and you can also consider porting by hand too, to be honest; it's a pretty standard approach)
- by the time you got this far, you probably already have a good idea of what you want to attempt next, and how you will do that :-)

**Comment by delijati** [comment]
Thanks for your quick response. I saw a glimpse of what you mean by trying to porting https://github.com/Cysu/cuda-kernel-benchmark ;)

**Comment by hughperkins** [comment]
Thats probably not a bad place to start, since it's small and self-contained.  You can try it with different compilers/runtimes, and see what works for you.  On the whole, if I was starting from scratch, I might be tempted to look at trisycl closely-ish, since Ronan is cool, trisycl is standards-based, and by the time stuff is working, SPIR-V cards are going to be relatively available, plausibly.

The main selling-point of cuda-on-cl is that it should work with almost any cards today, but it's a "bit" hacky, and SPIR-V is the better solution, for any cards/drivers which support it.

Well... there are actually two axes that can be varied:

| Input | Backend | Implementation |
|-------|-------|---|
| NVIDIA® CUDA™ | OpenCL 1.2 | cuda-on-cl |
| SYCL | SPIR-V | triSYCL (per my understanding) |
| NVIDIA® CUDA™ | SPIR-V | None, but might be a tempting option? |
| SYCL | OpenCL 1.2 | None, but not really tempting :-P |

**Comment by hughperkins** [comment]
See my comments at https://github.com/pytorch/pytorch/issues/488#issuecomment-294314810  This gives a plausible way to start.

**Comment by hughperkins** [comment]
Technically, this issue is done, since I have docuemnted, at a high-level, the steps required :-P
